iroh/src/baomap/flat.rs

```diff
@@ -0,0 +1,1210 @@
+    }
+
+    fn import_bytes(&self, data: Bytes) -> BoxFuture<'_, io::Result<Hash>> {
+        let this = self.clone();
+        tokio::task::spawn_blocking(move || this.import_bytes_sync(data))
+            .map(flatten_to_io)
+            .boxed()
+    }
+}
+
+impl State {
+    /// Gets or creates the outboard data for the given hash.
+    ///
+    /// For small entries the outboard consists of just the le encoded size,
+    /// so we create it on demand.
+    fn load_outboard(&self, size: u64, hash: &Hash) -> Option<Bytes> {
+        if needs_outboard(size) {
+            self.outboard.get(hash).cloned()
+        } else {
+            Some(Bytes::from(size.to_le_bytes().to_vec()))
+        }
+    }
+}
+
+impl Store {
+    fn import_sync(
+        self,
+        path: PathBuf,
+        mode: ImportMode,
+        progress: impl ProgressSender<Msg = ImportProgress> + IdGenerator,
+    ) -> io::Result<(Hash, u64)> {
+        if !path.is_absolute() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "path must be absolute",
+            ));
+        }
+        if !path.is_file() && !path.is_symlink() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "path is not a file or symlink",
+            ));
+        }
+        let id = progress.new_id();
+        progress.blocking_send(ImportProgress::Found {
+            id,
+            path: path.clone(),
+        })?;
+        let (hash, new, outboard) = match mode {
+            ImportMode::Reference => {
+                // compute outboard and hash from the data in place, since we assume that it is stable
+                let size = path.metadata()?.len();
+                progress.blocking_send(ImportProgress::Size { id, size })?;
+                let progress2 = progress.clone();
+                let (hash, outboard) = compute_outboard(&path, size, move |offset| {
+                    Ok(progress2.try_send(ImportProgress::OutboardProgress { id, offset })?)
+                })?;
+                progress.blocking_send(ImportProgress::OutboardDone { id, hash })?;
+                (
+                    hash,
+                    CompleteEntry::new_external(size, path.clone()),
+                    outboard,
+                )
+            }
+            ImportMode::Copy => {
+                let uuid = rand::thread_rng().gen::<[u8; 16]>();
+                let temp_data_path = self
+                    .0
+                    .options
+                    .partial_path
+                    .join(format!("{}.temp", hex::encode(uuid)));
+                // copy the data, since it is not stable
+                progress.try_send(ImportProgress::CopyProgress { id, offset: 0 })?;
+                let size = std::fs::copy(&path, &temp_data_path)?;
+                // report the size only after the copy is done
+                progress.blocking_send(ImportProgress::Size { id, size })?;
```

Comment:
this will panic, if the sync method is not called inside a tokio runtime, when using the tokio progress sender. you might want to use `flume` instead of `tokio::mpsc` for the progress sender, to avoid this

Reply:
I think this is not the end of the world, given that we are always in tokio land. But I will see if writing a FlumeProgressSender is straightforward.